Applications/Spire/Source/Ui/RealSpinBox.cpp

```diff
@@ -0,0 +1,441 @@
+using namespace Spire;
+
+namespace {
+  const auto SHIFT_STEPS = 10;
+  const auto MOUSE_REPEAT_DELAY_MS = 20;
+  const auto DOWN_ARROW_ICON = ":/Icons/arrow-down.svg";
+  const auto DOWN_ARROW_HOVER_ICON = ":/Icons/arrow-down-hover.svg";
+  const auto DOWN_ARROW_DISABLED_ICON = ":/Icons/arrow-down-disabled.svg";
+  const auto UP_ARROW_ICON = ":/Icons/arrow-up.svg";
+  const auto UP_ARROW_HOVER_ICON = ":/Icons/arrow-up-hover.svg";
+  const auto UP_ARROW_DISABLED_ICON = ":/Icons/arrow-up-disabled.svg";
+
+  auto ARROW_WIDTH() {
+    static auto width = scale_width(8);
+    return width;
+  }
+
+  auto clamped_real(RealSpinBox::Real value, RealSpinBox::Real min,
+      RealSpinBox::Real max) {
+    if(value.compare(min) < 0) {
+      return min;
+    } else if(value.compare(max) > 0) {
+      return max;
+    }
+    return value;
+  }
+}
+
+RealSpinBox::RealSpinBox(Real value, QWidget* parent)
+    : QAbstractSpinBox(parent),
+      m_minimum(std::numeric_limits<long double>::lowest()),
+      m_maximum(std::numeric_limits<long double>::max()),
+      m_step(1),
+      m_last_valid_value(value),
+      m_has_first_click(false),
+      m_up_arrow_timer_id(-1),
+      m_down_arrow_timer_id(-1) {
+  connect(lineEdit(), &QLineEdit::textChanged, this,
+    &RealSpinBox::on_text_changed);
+  setContextMenuPolicy(Qt::NoContextMenu);
+  lineEdit()->setText(display_string(m_last_valid_value));
+  set_decimal_places(MAXIMUM_DECIMAL_PLACES);
+  lineEdit()->installEventFilter(this);
+}
+
+void RealSpinBox::changeEvent(QEvent* event) {
+  if(event->type() == QEvent::EnabledChange) {
+    update_stylesheet();
+  }
+  QAbstractSpinBox::changeEvent(event);
+}
+
+bool RealSpinBox::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == lineEdit() && event->type() == QEvent::MouseButtonPress &&
+      !m_has_first_click && isEnabled()) {
+    auto e = static_cast<QMouseEvent*>(event);
+    if(e->button() == Qt::LeftButton) {
+      m_has_first_click = true;
+      selectAll();
+      return true;
+    }
+  }
+  return QAbstractSpinBox::eventFilter(watched, event);
+}
+
+void RealSpinBox::focusInEvent(QFocusEvent* event) {
+  if(event->reason() != Qt::MouseFocusReason) {
+    m_has_first_click = true;
+  }
+  QAbstractSpinBox::focusInEvent(event);
+}
+
+void RealSpinBox::focusOutEvent(QFocusEvent* event) {
+  m_has_first_click = false;
+  blockSignals(true);
+  QAbstractSpinBox::focusOutEvent(event);
+  blockSignals(false);
+  on_editing_finished();
+}
+
+void RealSpinBox::keyPressEvent(QKeyEvent* event) {
+  if(event->modifiers().testFlag(Qt::ControlModifier)) {
+    QAbstractSpinBox::keyPressEvent(event);
+    return;
+  }
+  switch(event->key()) {
+    case Qt::Key_Backspace:
+      lineEdit()->backspace();
+      return;
+    case Qt::Key_Delete:
+      lineEdit()->setText(display_string(m_last_valid_value));
+      return;
+    case Qt::Key_Enter:
+    case Qt::Key_Return:
+      on_editing_finished();
+      return;
+    case Qt::Key_PageDown:
```

Comment:
What's the purpose of case ```Key_PageDown/Up```, is it the case that the ```event->text()``` is not empty on those two key presses?

Reply:
If the QAbstractSpinBox gets the page up/down key event, it ultimately calls its own implementation of stepBy(...) that selects/highlights all of the input text.

If you want the page up/down to step the input value I can just have it call the stepBy(...) function I defined instead of returning.

Reply:
Nevermind, that was true before but the condition following the switch makes this impossible.